runtime/j9vm/jvm.c

```diff
@@ -1717,12 +1717,57 @@ JNI_CreateJavaVM_impl(JavaVM **pvm, void **penv, void *vm_args, BOOLEAN isJITSer
 	 * trying to load application native libraries that are linked against
 	 * libraries in /usr/lib we could fail to find those libraries if /usr/lib
 	 * is not on the LIBPATH.
-	 * */
-	addToLibpath("/usr/lib", FALSE);
+	 */
+	{
+		/* github #8504 shows tests fail if the libpath is modified
+		 * when creating a new process from Java.  The easy fix is to only
+		 * append to the libpath if /usr/lib isn't already present.
+		 * Example libpath:
+		 * LIBPATH=/jre/lib/ppc64/j9vm:/jre/lib/ppc64:/jre/lib/ppc64/jli:/jre/../lib/ppc64:/usr/lib
+		 */
+
+		const char *currentLibPath = getenv("LIBPATH");
+		BOOLEAN appendToLibPath = TRUE;
+		if (NULL != currentLibPath) {
+			const size_t currentLibPathLength = strlen(currentLibPath);
+			const char *usrLib = "/usr/lib";
+			const UDATA usrLibLength = LITERAL_STRLEN("/usr/lib");
+			const char *needle = strstr(currentLibPath, usrLib);
+			while (NULL != needle) {
+				/* Note, inside the loop we're guaranteed to have 
+				 * usrLibLength of string to operate on.
+				 */
+				const ptrdiff_t offsetFromStart = needle - currentLibPath;
+				const size_t remainingLength = currentLibPathLength - offsetFromStart;
+				if (0 != offsetFromStart) {
+					if (needle[-1] != ':') {
+						goto nextCheck;
+					}
+				}
+				if (remainingLength == usrLibLength) {
+					/* Found a match */
+					appendToLibPath = FALSE;
+					break;
+				} else if (remainingLength > usrLibLength) {
+					if (needle[usrLibLength] == ':') {
+						/* Found a match */
+						appendToLibPath = FALSE;
+						break;
+					}
+				}
```

Comment:
This can be simplified:
* We know `remainingLength >= usrLibLength` else `strstr()` would have returned `NULL`.
* It's safe to peek at the character at `needle[usrLibLength]` unconditionally (at worst it's the NUL terminator).
```
    if ((':' == needle[usrLibLength]) || ('\0' == needle[usrLibLength])) {
        /* Found a match */
        appendToLibPath = FALSE;
        break;
    }
```
Nesting it within the inverse of the `if` statement above we can remove the goto and the label:
```
    if ((0 == offsetFromStart) || (':' == needle[-1])) {
        if ((':' == needle[usrLibLength]) || ('\0' == needle[usrLibLength])) {
            /* Found a match */
            appendToLibPath = FALSE;
            break;
        }
    }                   
```